### PR TITLE
MAINT: array types: restrict to boolean & numerical dtypes

### LINKED ISSUE #19276: MAINT: array types: make `compliance_scipy` more strict
Following gh-19187, there is desire to restrict the types of array we accept. I'm collecting the relevant comments here for future reference when a PR wants to deal with this. A test in gh-19260 which involves a string dtype is tripping us up.

@rgommers:

> I'd say it should be only boolean and numerical dtypes (signed/unsigned integer and real/complex floating-point dtypes).

@mdhaber:

> I think we can restrict to [subdtypes](https://numpy.org/doc/stable/reference/arrays.scalars.html) of `np.number` and `np.bool_`.

> are we going to be bothered by the overhead of np.issubdtype? It's probably the slowest thing in this function, and in the simplest implementation, it would be called twice per argument.

```python
import numpy as np
x = np.asanyarray(['a'])
%timeit np.issubdtype(x.dtype, np.flexible)  # 646 ns ± 2.71 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

@rgommers:

> The overhead of ~650 ns is quite a bit, so we may have to optimize this later. But it's not a big worry for now - we can circle back to that once it start showing up in profiling results.

---

Related to this is the possibility of optional parameters in `array_namespace` and `compliance_scipy`, or `as_xparray`, which could eventually eliminate the need for [`scipy._lib._util._asarray_validated`](https://github.com/scipy/scipy/blob/38410650cc0b0ffb058ec57b154a0bd688db8b8d/scipy/_lib/_util.py#L240-L293) (used in `linalg`). Here are its optional parameters:

```python
"""
...
check_finite : bool, optional
    Whether to check that the input matrices contain only finite numbers.
    Disabling may give a performance gain, but may result in problems
    (crashes, non-termination) if the inputs do contain infinities or NaNs.
    Default: True
sparse_ok : bool, optional
    True if scipy sparse matrices are allowed.
objects_ok : bool, optional
    True if arrays with dype('O') are allowed.
mask_ok : bool, optional
    True if masked arrays are allowed.
as_inexact : bool, optional
    True to convert the input array to a np.inexact dtype.
...
"""
```

### PR DESCRIPTION
**Pull request by lucascolley**
#### Reference issue
Closes gh-19276

#### What does this implement/fix?
- Restrict input dtypes to subdtypes of `np.number` and `np.bool_` when the `SCIPY_ARRAY_API` flag is set

#### Additional information
The other points in gh-19276 can be moved to gh-18972 as part of the wider validation overhaul.

cc @mdhaber @rgommers

### CODE REVIEW

**Review by mdhaber — APPROVED** [review]
I think this looks good if Cirrus is happy. I'd just capitalize NumPy where it's used as a word.

File: scipy/_lib/_array_api.py [hunk 1]
```diff
@@ -35,29 +35,37 @@ def compliance_scipy(arrays):
...
 
     The following subclasses are not supported and raise and error:
-    - `np.ma.MaskedArray`
+    - `numpy.ma.MaskedArray`
     - `numpy.matrix`
-    - Any array-like which is not Array API compatible or coercible by numpy
-    - object arrays
+    - numpy arrays which do not have a boolean or numerical dtype
+    - Any array-like which is neither array API compatible nor coercible by numpy
+    - Any array-like which is coerced by numpy to an unsupported dtype
```
Comment (lucascolley) [inline scipy/_lib/_array_api.py:42]:
```suggestion
    - NumPy arrays which do not have a boolean or numerical dtype
    - Any array-like which is neither array API compatible nor coercible by NumPy
    - Any array-like which is coerced by NumPy to an unsupported dtype
```

File: scipy/_lib/_array_api.py [hunk 2]
```diff
@@ -35,29 +35,37 @@ def compliance_scipy(arrays):
...
             try:
                 array = np.asanyarray(array)
             except TypeError:
-                raise TypeError("Array is not Array API compatible or "
-                                "coercible by numpy")
-            if array.dtype is np.dtype('O'):
-                raise TypeError("An argument was coerced to an object array, "
-                                "but object arrays are not supported.")
+                raise TypeError("An argument is neither array API compatible nor "
+                                "coercible by numpy.")
```
Comment (lucascolley) [inline scipy/_lib/_array_api.py:60]:
```suggestion
                                "coercible by NumPy.")
```

**Comment by lucascolley** [comment]
> I think this looks good if Cirrus is happy. I'd just capitalize NumPy where it's used as a word.

Cirrus is happy. NumPy change made (force push to skip ci).

File: scipy/_lib/_array_api.py [hunk 3]
```diff
@@ -35,29 +35,37 @@ def compliance_scipy(arrays):
...
+    - Any array-like which is coerced by NumPy to an unsupported dtype
     """
     for i in range(len(arrays)):
         array = arrays[i]
         if isinstance(array, np.ma.MaskedArray):
-            raise TypeError("'numpy.ma.MaskedArray' are not supported")
+            raise TypeError("'numpy.ma.MaskedArray' arrays are not supported.")
         elif isinstance(array, np.matrix):
-            raise TypeError("'numpy.matrix' are not supported")
+            raise TypeError("'numpy.matrix' arrays are not supported.")
```
Comment (lucascolley) [inline scipy/_lib/_array_api.py:49]: this probably should not say `arrays` here, but I don't like how it sounds without it - `matrices` instead?

Comment (mdhaber) [reply scipy/_lib/_array_api.py:49]: I didn't think so either, but chose not to nit : ) 

"Inputs of type `numpy.matrix` are not supported." or "`numpy.matrix` objects are not supported."

**Comment by lucascolley** [comment]
Wording nits addressed 👍 (please squash merge, I've been lazy with commit messages)

**Comment by mdhaber** [comment]
Thanks @lucascolley. Please do add those additional thoughts from gh-19276 to gh-18972.

**Pull request merged**